### PR TITLE
feat: Screenshots

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
The Screenshot feature required changes in `sentry-java` and `sentry-cocoa`. The `sentry-react-native` uses the native SDKs methods to capture screenshots. Those are copied over the bridge and the envelope bytes including the screenshots are assembled on the JS/RN side and then passed back to the native layers to be sent.

In case of native crashes, native SDKs capture screenshots themselves. 

- https://github.com/getsentry/sentry-java/pull/2360
	- Waits for @romtsn to be released (Likely Wednesday 23. Nov 2022)
- https://github.com/getsentry/sentry-cocoa/pull/2384
	- Released in `7.31.1` 	

The screenshots are captured in the `client` and not in the integration due to Async/Await JS nature. In the case of multiple async integrations, the screen could be different if captured in integration.

## :bulb: Motivation and Context
Closes: https://github.com/getsentry/sentry-react-native/pull/2373
Closes: https://github.com/getsentry/sentry-react-native/issues/2362

## :green_heart: How did you test it?
Using sample app.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

- [ ] Merge the new `sentry-jave` to main rebase this PR and if CI passes merge.

### CODE REVIEW

**Comment by marandaneto** [comment]
Consider https://github.com/getsentry/sentry-react-native/issues/2362#issuecomment-1193763553
Only this is not enough.

**Comment by krystofwoldrich** [comment]
> Consider [#2362 (comment)](https://github.com/getsentry/sentry-react-native/issues/2362#issuecomment-1193763553) Only this is not enough.

I know I've read it.

File: android/src/main/java/io/sentry/react/RNSentryModule.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModule.java:305]: Consider this check as well https://github.com/getsentry/sentry-java/blob/38e4f116623e0dd1329f54d86dd4bbe5dd687a86/sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java#L182-L186

File: android/src/main/java/io/sentry/react/RNSentryModule.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModule.java:304]: Maybe we could refactor `ScreenshotEventProcessor` on android that takes an `Activity` as a parma and reuse this class/method in here, so we avoid duplication and increase of bundle size.

File: ios/RNSentry.m
Comment (marandaneto) [inline ios/RNSentry.m:294]: M: Can we actually reuse this? https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentryScreenshot.m

File: android/src/main/java/io/sentry/react/RNSentryModule.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModule.java:139]: H: This alone won't work for Native events, See https://github.com/getsentry/sentry-react-native/issues/2362#issuecomment-1193763553
Maybe we should expose a `ScreenshotEventProcessor#setCurrentActivity(activity)` so the RN SDK can overwrite this value when initing the Android SDK.

File: ios/RNSentry.m
Comment (krystofwoldrich) [reply ios/RNSentry.m:294]: I think we could expose `takeScreenshots` using the PrivateSDKs header.

File: src/js/wrapper.ts
Comment (marandaneto) [inline src/js/wrapper.ts:470]: L: Should we log it?

File: src/js/client.ts
Comment (marandaneto) [inline src/js/client.ts:212]: M: Does it make sense to be an integration or event processor instead?
If not, that's ok, but something we have to do is to at least add an integration in the `sdk.integrations` list so we can track adoption of this feature.
Only if `attachScreenshot` is enabled of course.

Comment (krystofwoldrich) [reply src/js/client.ts:212]: As the order of the integrations is not guaranteed, and integrations can be async the screen could be captured too late. Or would that still be good enough? I know other platforms also use the event processor for the screenshots.

I'll add the Integration so we can track the adoption.

**Comment by marandaneto** [comment]
@krystofwoldrich is this ready for a final review os is there something missing?
The PR description is not fully filled.

**Comment by krystofwoldrich** [comment]
@marandaneto Ready for final review, it only waits for `sentry-java` to be released. I talked to @romtsn and he would like to finish something and then release it (likely this Wednesday).

File: android/src/main/java/io/sentry/react/RNSentryModule.java
Comment (marandaneto) [inline android/src/main/java/io/sentry/react/RNSentryModule.java:313]: Can you call `getCurrentActivity()` directly here?

File: ios/RNSentry.m
Comment (marandaneto) [inline ios/RNSentry.m:302]: M: Can this be `nil` or 0 bytes or something? we do checks on Android but not on iOS?

**Comment by marandaneto** [comment]
@krystofwoldrich Since you are blocked by the Java release, we can draft up the docs PR for RN, similar to https://docs.sentry.io/platforms/flutter/enriching-events/screenshots/

**Review by marandaneto — APPROVED** [review]

File: ios/RNSentry.m
Comment (krystofwoldrich) [reply ios/RNSentry.m:302]: I don't know, but I would say, if it can, we should check it in `cocoa` and not add such a screenshot to the array. 

@brustolin Do you know, if [this](https://github.com/getsentry/sentry-cocoa/blob/f781ecba1e2ccfc37e2156f1a9300452c0c8f4c8/Sources/Sentry/SentryScreenshot.m#LL49C17-L49C17) can be `nil` sometimes or 0 size?

File: android/src/main/java/io/sentry/react/RNSentryModule.java
Comment (krystofwoldrich) [reply android/src/main/java/io/sentry/react/RNSentryModule.java:313]: I can, thanks.

File: ios/RNSentry.m
Comment (brustolin) [reply ios/RNSentry.m:302]: Theoretically no. And I never saw it happening. 
Nil here is not a problem, the loop will ignore nil, but I guess you could add a check for 0 size.

Comment (krystofwoldrich) [reply ios/RNSentry.m:302]: @brustolin Should I add the 0 size check in the cocoa SDK, I guess the 0 size screenshot should not be in the list.

Comment (brustolin) [reply ios/RNSentry.m:302]: Good point. That would be the best solution

**Comment by marandaneto** [comment]
@krystofwoldrich can we unblock this? it should not be a problem cutting a patch release on Android.

**Comment by krystofwoldrich** [comment]
@marandaneto The release is out. So this is not blocked anymore.

**Comment by marandaneto** [comment]
> @marandaneto The release is out. So this is not blocked anymore.

I'd way for this https://github.com/getsentry/publish/issues/1614
The release 6.9.0 isn't fully finished, not sure if it's reliable yet.

File: ios/RNSentry.m
Comment (krystofwoldrich) [reply ios/RNSentry.m:312]: - https://github.com/getsentry/sentry-cocoa/pull/2459

**Comment by marandaneto** [comment]
> > @marandaneto The release is out. So this is not blocked anymore.
> 
> I'd way wait this [getsentry/publish#1614](https://github.com/getsentry/publish/issues/1614) The release 6.9.0 isn't fully finished, not sure if it's reliable yet.

@krystofwoldrich you can unblock this -> https://github.com/getsentry/sentry-react-native/pull/2653

File: CHANGELOG.md
Comment (marandaneto) [inline CHANGELOG.md:5]: @krystofwoldrich make it to `Unreleased`

Comment (krystofwoldrich) [reply CHANGELOG.md:5]: On it. :)

**Pull request merged**
